### PR TITLE
XVS Vault

### PR DESCRIPTION
**Pull request by ferrari105**
## Description

As the XVS vault branch, this has following features:
- Implemented multiple token reward feature thought Store contract, e.g by depositing XVS, user can get rewarded XVS.
Reward token distribution is controlled by `rewardTokenPerBlocks` and `allocPoint` of each pool. Reward token amount per block is different per reward token. 
- Implemented withdrawal unlock period and in order to withdraw, user needs to request withdrawal with`requestWithdrawal()`. After lock period is over, user can execute withdrawal by calling `executeWithdrawal()`. After user requests withdrawal once, user can request again after execute his request.
- Added `getPriorVotes` function for governor and it returns user's XVS deposited amount in xvs depositing pool.

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [x] If I added new functionality, I added tests covering it.
- [x] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: tests/XVSVault/xvsVaultProxyTest.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        expect(result).toHaveTrollFailure('UNAUTHORIZED', 'ACCEPT_PENDING_IMPLEMENTATION_ADDRESS_CHECK');
+      });
+
+      it("does not change current implementation address", async () => {
+        expect(await call(vaultProxy, 'xvsVaultImplementation')).not.toEqual(vaultProxy._address);
+      });
+    });
+
+    it.skip("rejects if pending impl is address(0)", async () => {
+      // XXX TODO?
```
Comment (kkirka) [inline tests/XVSVault/xvsVaultProxyTest.js:75]: Are you going to add this test?

File: tests/XVSVault/xvsVaultProxyTest.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        // Does our log decoder expect it to come from the same contract?
+        // Having difficulty decoding these events
+        // assert.toHaveLog(
+        //   result,
+        //   "NewPendingImplementation",
+        //   {
+        //     oldPendingImplementation: vaultImpl._address,
+        //     newPendingImplementation: "0x0000000000000000000000000000000000000000"
+        //   });
+      });
```
Comment (kkirka) [inline tests/XVSVault/xvsVaultProxyTest.js:117]: Same for these two — do you plan to complete them?

File: contracts/XVSVault/XVSVault.sol [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+        uint256 length = poolInfo.length;
+        for (uint256 pid = 0; pid < length; ++pid) {
+            if (address(poolInfo[pid].token) == address(xvsAddress)) {
+                UserInfo storage user = userInfos[xvsAddress][pid][account];
+                return uint96(user.amount);
+            }
+        }
+        return uint96(0);
+    }
```
Comment (kkirka) [inline contracts/XVSVault/XVSVault.sol:320]: This function just returns the user's XVS balance in the vault, not the voting power, is this supposed to be changed in the governance PR?

File: contracts/XVSVault/XVSVault.sol [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+        user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
+        emit Deposit(msg.sender, _rewardToken, _pid, _amount);
+    }
+
+    /**
+     * @notice Execute withdrawal to XVSVault for XVS allocation
+     * @param _rewardToken The Reward Token Address
+     * @param _pid The Pool Index
+     */
+    function ExecuteWithdrawal(address _rewardToken, uint256 _pid) public nonReentrant {
```
Comment (kkirka) [inline contracts/XVSVault/XVSVault.sol:155]: Why does it start with an uppercase letter though, any particular reason for this?

```suggestion
    function executeWithdrawal(address _rewardToken, uint256 _pid) public nonReentrant {
```

File: contracts/XVSVault/XVSVault.sol [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        emit ExecutedWithdraw(msg.sender, _rewardToken, _pid, _amount);
+    }
+
+    /**
+     * @notice Request withdrawal to XVSVault for XVS allocation
+     * @param _rewardToken The Reward Token Address
+     * @param _pid The Pool Index
+     * @param _amount The amount to withdraw to vault
+     */
+    function RequestWithdrawal(address _rewardToken, uint256 _pid, uint256 _amount) public nonReentrant {
```
Comment (kkirka) [inline contracts/XVSVault/XVSVault.sol:186]: Same here

```suggestion
    function requestWithdrawal(address _rewardToken, uint256 _pid, uint256 _amount) public nonReentrant {
```

**Review by kkirka — CHANGES_REQUESTED** [review]
I left several comments there, these should be addressed. I'll make another pass a bit later.

File: contracts/XVSVault/XVSVault.sol [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        uint256 pending =
+            user.amount.mul(pool.accRewardPerShare).div(1e12).sub(
+                user.rewardDebt
+            );
+        IXVSStore(xvsStore).safeRewardTransfer(_rewardToken, msg.sender, pending);
+        user.amount = user.amount.sub(_amount);
+        user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
+        pool.token.transfer(address(msg.sender), _amount);
+
+        withdrawal.amount = 0;
```
Comment (gryincent) [inline contracts/XVSVault/XVSVault.sol:175]: @ferrari105 please use this statement before token transfer.

File: contracts/XVSVault/XVSVault.sol [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+        require(lockPeriod.add(withdrawal.timestamp) < curTimestamp, "your request is locked yet");
+
+        updatePool(_rewardToken, _pid);
+        uint256 pending =
+            user.amount.mul(pool.accRewardPerShare).div(1e12).sub(
+                user.rewardDebt
+            );
+        IXVSStore(xvsStore).safeRewardTransfer(_rewardToken, msg.sender, pending);
+        user.amount = user.amount.sub(_amount);
+        user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
```
Comment (gryincent) [inline contracts/XVSVault/XVSVault.sol:172]: we need to update the calculation before reward transfer. @ferrari105

File: contracts/XVSVault/XVSVault.sol [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+     * @param _rewardToken The Reward Token Address
+     * @param _pid The Pool Index
+     * @param _amount The amount to withdraw to vault
+     */
+    function requestWithdrawal(address _rewardToken, uint256 _pid, uint256 _amount) public nonReentrant {
+        UserInfo storage user = userInfos[_rewardToken][_pid][msg.sender];
+        WithdrawalInfo storage withdrawal = withdrawalInfos[_rewardToken][_pid][msg.sender];
+        require(_amount > 0, "requested amount cant be zero");
+        require(user.amount >= _amount, "requested amount is invalid");
+        require(withdrawal.amount == 0, "request again after execute");
```
Comment (gryincent) [inline contracts/XVSVault/XVSVault.sol:191]: users can make several withdrawal request. we need to update some workflows and storages to implement this. @ferrari105

**Review by gryincent — COMMENTED** [review]
1. users can make several withdrawal request.
2. we should use storage update commands before real token transfer, although there is reentrancy check.
@ferrari105 please consider these changes.

**Review by kkirka — APPROVED** [review]

**Pull request merged**